examples/with-cookie-auth/utils/auth.js

```diff
@@ -0,0 +1,82 @@
+import { Component } from 'react'
+import Router from 'next/router'
+import nextCookie from 'next-cookies'
+import cookie from 'js-cookie'
+import fetch from 'isomorphic-unfetch'
+
+export const login = async ({ username }) => {
+  try {
+    const url = 'https://with-cookie-api.now.sh/login'
+    const response = await fetch(url, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify({ username })
+    })
+    if (response.ok) {
+      const { token } = await response.json()
+      cookie.set('token', token, { expires: 1 })
```

Comment:
Is this a good example? Storing access tokens without the `secure` and `httpOnly` attributes is vulnerable to XSS attacks. If we want to omit `httpOnly`, I'd suggest at least adding some info to the readme. What do you think?

Reply:
You're right, but I tried to do the example as simple as possible. The only thing I'm trying to demo here is the minimal implementation you can do for cookie auth, the security and other concerns are out of the scope just for this demo.

I totally have to state this in the `readme.md`.

Reply:
Yeah, it would be awesome to state that in there. People often look at the examples as "best practices" - this is my only concern here 😄

Reply:
@dav-is Of course it works, you just have to do it differently. As simple as an endpoint which returns user data if you have a valid token (cookie) does the job here. But of course we don't want to do anything like this in this example - that's why I suggested adding some info in the README.

Reply:
@sarneeh User logs in from client device, API adds httpOnly cookie to it's response. Fetch results from API, the browser automatically keeps the cookie in the request. The user does a hard refresh and SSR kicks in. The SSR fetches data from the API (except now it doesn't have a cookie because the other is stored on the browser of the other domain).

I guess what you're saying could be possible if the API was in a subdirectory, but I like separating SSR and the API by a subdomain. I personally don't like using cookies for auth in an API. If you're using an Auth header with a token appended on each request makes CSRF nearly impossible. Also something about using a cookie seems stateful and doesn't fit with a stateless API.

Reply:
@dav-is I don't understand. Why do you assume that the cookie is for a different domain? This doesn't make sense to me. When you have an API endpoint for logging in, I think it's logical that it sets a cookie for the domain where the frontend is on. You can have your Next.js server on `domain.com` and the auth API on `api.domain.com` and it just works.

Reply:
Would `www.domain.tld` to `api.domain.tld` work? When I store cookies I also store them with the current domain written out so a vulnerability on another subdomain couldn't leak a users cookie.

And the difference it makes in XSS is minimal. If you had an XSS vulnerability and you were using `httpOnly` the malicious script might not be able to steal your token, but it could just make requests on your behalf. An XSS vulnerability could just get your username and password plaintext when a user logs in—at that point who cares about the token...

Being able to just not worry about CSRF is great and it instantly makes your API more easily made public. If you had a client that allowed you to log into multiple accounts at once and switch accounts and just provide a different Auth header with each request. This way you could maintain each session just by storing the token. Cookies are weird and stateful I kind of hate the idea of them with an API lol

Reply:
@dav-is Huh, you're right! You kinda persuaded me against cookies now 😄 Even using `httpOnly`, when I'm vulnerable to XSS then I'm... busted. I don't know why I didn't come into this realization earlier, thanks! 😄